rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal increases the size of proofs, and increases the number of checks
+that the verifier must perform, and thus the cost of the recusive verifier.
+This also increases the complexity of the proof system.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Why is this design the best in the space of possible designs?
+
+This design combines the 3 different 'modes' of random access memory, roughly
+approximating `ROM -> RAM` loading, `disk -> RAM` loading, and `RAM r/w`. This
+gives the maximum flexibility for circuit designers, and each primitive has
+uses for projects that we are persuing.
+
+This design has been refined over several iterations to reduce the number of
+components exposed in the proof and the amount of computation that the verifier
+requires. The largest remaining parts represent the data itself and the method
+for combining them, where simplifying assumptions have been made to minimize
+their impact while still satisfying the design constraints.
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+Other designs building upon the lookup primitive reduce on some subset of this
+design, or on relaxing a design constraint that would remove some useful /
+required functionality.
+
+### What is the impact of not doing this?
+
+We are currently unable to expose certain useful functionality to users of
+Snapps / SnarkyJS, and are missing the ability to use that same functionality
+for important parts of the zkOracles ecosystem (such as parsing / scraping).
+
+We also are currently forced to hard-code any constraints that we need for
+particular uses in the verifier circuit, and users are unable to remove the
+parts they do not need and to replace them with other useful constraints that
+their circuits might benefit from.
+
+The Mina transaction model is unable to support a bytecode interpreter without
+this or similar extensions.
+
+## Prior art
+[prior-art]: #prior-art
+
+*(Elided)*
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+### What parts of the design do you expect to resolve through the RFC process before this gets merged?
+
+* The maximum number of tables / entries of each kind that we want to support in the 'standard' protocol.
+* Validation (or rejection) of the simplifying assumptions.
+* Confirmation of the design constraints, or simplifications thereto.
+* Necessity / utility of the permutation argument.
+
+### What parts of the design do you expect to resolve through the implementation of this feature before merge?
+
+None, this RFC has been developed in part from an initial design and iteration
+on implementations of parts of that design.
+
+### What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+
```

Comment:
What design choices for this system or it's parameters could be impacted by the design of these future use-cases?